content/blog/posts/2022-08-04-add-desktop-notifications.md

```diff
@@ -0,0 +1,241 @@
+
+```js
+async _showNotification(worker) {
+  const permission = await this._isShowNotificationPermitted();
+  if (permission !== 'denied') {
+    worker.showNotification('Be notified of new posts', {
+      body: 'Let us notifiy you of new posts on our site!',
+      requireInteraction: true,
+      actions: [
+        {
+          action: 'dismiss',
+          title: 'Dismiss'
+        },
+        {
+          action: 'subscribe',
+          title: 'Subscribe'
+        }
+      ]
+    });
+  }
+}
+```
+
+The difference now is that we can set actions on the notification options and allow the user to select something from that notification. This click will get passed to the worker and given what the action is, we can react and do something. So, let's do just that and act on the user's response, which we're requiring with `requireInteraction`.
+
+```js
+self.addEventListener('notificationclick', function(event) {
+  const { action, notification } = event;
+  notification.close();
+  
+  if (action === 'subscribe') {
+    const promiseChain = clients.openWindow('https://feedly.com/i/subscription/feed%2Fhttps%3A%2F%2Fshipshape.io%2Ffeed.xml');
+    event.waitUntil(promiseChain);
+  }
+});
+```
+
+Here I've updated the worker to check for an action if the user click's `subscribe` button in the notification, when it's captured and checked we open a new window with out RSS subscription link. If they click `dismiss`, nothing happens and the notification is dismissed. In case you want some kind of persistence, we could add some way to track the user selecting to dismiss and save that state to something like localStorage.
+
+![Desktop notification with actions](/img/blog/add-desktop-notifications/notification-with-actions.png)
+
+In the end, we have a class to setup our service worker and show a desktop notification to the user to allow them to subscribe to our blog posts. Here's what it looks like all together. 
+
+```js
+class NotificationService {
+  constructor() {
+    this._startServiceWorker(this);
+  }
+
+  _startServiceWorker(notificationService) {
+    if ('serviceWorker' in navigator) {
+      navigator.serviceWorker.register('/landing/js/service-worker.js')
+      .then(function(registration) {
+        registration.addEventListener('updatefound', function() {
+          const installingWorker = registration.installing;
+          console.log('A new service worker is being installed:', installingWorker);
+        });
+
+        notificationService._showNotification(registration);
+      })
+      .catch(function(error) {
+        console.log('Service worker registration failed:', error);
+      });
+    } else {
+      console.log('Service workers are not supported.');
+    }
+  }
+
+  async _isShowNotificationPermitted() {
+    if ('Notification' in window) {
+      let { permission } = Notification;
+      if (permission !== 'granted') {
```

Comment:
Will this re-request the permission if it's already been denied?

Reply:
ah yes, so this would be better

```
if (permission === 'granted') {
        return true;
      }
      if (permission !== 'denied') {
        permission =  await Notification.requestPermission();
      }
```